node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Prunable queue mimics a producer-consumer channel, in enabling
+//! waiting on the receiver side for new items, while providing pruning functionality.
+//! The pruning takes place whenever a new item is sent, based on a specified predicate.
+//!
+//! The queue facilitates the asynchronous return of processing result for each added item.
+//!
+//! The separation of `Sender` and `Receiver` is employed primarily because `Receiver` requires
+//! a mutable reference to the signaling channel, unlike `Sender`, hence making it undesirable to
+//! be used in conjunction.
+
+use crate::{
+    ctx::{self, Ctx},
+    oneshot,
+    sync::{self, watch},
+};
+use std::{collections::VecDeque, sync::Arc};
+use tokio::sync::Mutex;
+
+pub fn new<T, U>(
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+) -> (Sender<T, U>, Receiver<T, U>) {
+    let queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>> = Mutex::new(VecDeque::new());
+    // Internal signaling, to enable waiting on the receiver side for new items.
+    let (send, recv) = watch::channel(false);
+
+    let shared = Arc::new(Shared {
+        queue,
+        has_items_send: send,
+    });
+
+    let sender = Sender {
+        shared: shared.clone(),
+        pruning_predicate: pruning_predicate,
+    };
+
+    let receiver = Receiver {
+        shared: shared.clone(),
+        has_items_recv: recv,
+    };
+
+    return (sender, receiver);
+}
+
+pub struct Shared<T, U> {
+    queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>>,
+    has_items_send: watch::Sender<bool>,
+}
+
+pub struct Sender<T, U> {
+    shared: Arc<Shared<T, U>>,
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+}
+
+impl<T, U> Sender<T, U> {
+    pub async fn enqueue(&self, item: T) -> oneshot::Receiver<U> {
+        // Create oneshot channel for returning result asynchronously.
+        let (res_send, res_recv) = oneshot::channel();
+
+        let mut queue = self.shared.queue.lock().await;
+        queue.retain(|existing_item| !(self.pruning_predicate)(&existing_item.0, &item));
```

Comment:
as far as I'm concerned, with our use case we are good with running the predicate over all messages in the queue when a new message arrives, but please remember to document that pushing a new message takes O(n).

Reply:
Fixed.